src/encode.js

```diff
@@ -9,85 +9,117 @@ import ParsePolygon from './ParsePolygon';
 
 function encode(
   value: mixed,
   disallowObjects: boolean,
   forcePointers: boolean,
   seen: Array<mixed>,
-  offline: boolean
+  offline: boolean,
+  counter: number,
+  initialValue: mixed
 ): any {
+  counter++;
+
+  if (counter > MAX_RECURSIVE_CALLS) {
+    const message = 'Encoding object failed due to high number of recursive calls, likely caused by circular reference within object.';
+    console.error(message);
+    console.error('Value causing potential infinite recursion:', initialValue);
+
+    throw new Error(message);
+  }
+
   if (value instanceof ParseObject) {
     if (disallowObjects) {
       throw new Error('Parse Objects not allowed here');
     }
-    const seenEntry = value.id ? value.className + ':' + value.id : value;
+    const entryIdentifier = value.id ? value.className + ':' + value.id : value;
     if (
       forcePointers ||
-      !seen ||
-      seen.indexOf(seenEntry) > -1 ||
+      seen.includes(entryIdentifier) ||
       value.dirty() ||
-      Object.keys(value._getServerData()).length < 1
+      Object.keys(value._getServerData()).length === 0
     ) {
       if (offline && value._getId().startsWith('local')) {
         return value.toOfflinePointer();
       }
       return value.toPointer();
     }
-    seen = seen.concat(seenEntry);
+    seen.push(entryIdentifier);
     return value._toFullJSON(seen, offline);
-  }
-  if (
+  } else if (
     value instanceof Op ||
     value instanceof ParseACL ||
     value instanceof ParseGeoPoint ||
     value instanceof ParsePolygon ||
     value instanceof ParseRelation
   ) {
     return value.toJSON();
-  }
-  if (value instanceof ParseFile) {
+  } else if (value instanceof ParseFile) {
     if (!value.url()) {
       throw new Error('Tried to encode an unsaved file.');
     }
     return value.toJSON();
-  }
-  if (Object.prototype.toString.call(value) === '[object Date]') {
+  } else if (Object.prototype.toString.call(value) === '[object Date]') {
     if (isNaN(value)) {
       throw new Error('Tried to encode an invalid date.');
     }
     return { __type: 'Date', iso: (value: any).toJSON() };
-  }
-  if (
+  } else if (
     Object.prototype.toString.call(value) === '[object RegExp]' &&
     typeof value.source === 'string'
   ) {
     return value.source;
-  }
-
-  if (Array.isArray(value)) {
-    return value.map(v => {
-      return encode(v, disallowObjects, forcePointers, seen, offline);
-    });
-  }
-
-  if (value && typeof value === 'object') {
+  } else if (Array.isArray(value)) {
+    return value.map(v => encode(v, disallowObjects, forcePointers, seen, offline, counter, initialValue));
+  } else if (value && typeof value === 'object') {
     const output = {};
     for (const k in value) {
-      output[k] = encode(value[k], disallowObjects, forcePointers, seen, offline);
+      try {
+        // Attempts to get the name of the object's constructor
+        // Ref: https://stackoverflow.com/a/332429/6456163
```

Comment:
Same topic as earlier. If you copy/pasted someone else's code from SO then we cannot accept it.